packages/app/src/lib/schema/primitives.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * Primitive types for the event schema.
+ * These are the building blocks used across all event definitions.
+ *
+ * Lexicon mapping notes:
+ * - ulid -> { type: 'string' } (no ulid format in lexicons)
+ * - hash -> { type: 'string' } (could use description to note it's hex)
+ * - did -> { type: 'string', format: 'did' }
+ * - cid -> { type: 'cid' } (lexicons have native cid support)
+ */
+
+import { type, Type } from "arktype";
+import { type Bytes as BytesLink } from "@atcute/cbor";
+import { isDid } from "@atproto/oauth-client";
+import { isValid as isValidUlid } from "ulidx";
+
+export type Bytes = BytesLink;
+export const Bytes = type({ $bytes: "string.base64" });
+
+export const ulid = type.string
+  .narrow((v, ctx) => (isValidUlid(v) ? true : ctx.mustBe("a valid ULID")))
+  .brand("ulid");
+export type Ulid = typeof ulid.infer;
+
+export const did = type.string
+  .narrow((v, ctx) => (isDid(v) ? true : ctx.mustBe("a valid DID")))
+  .brand("did");
+export type Did = typeof did.infer;
+
+// Hash: 64 hex characters (SHA-256)
+// In DRISL this is 32 bytes, decoded to hex string
+export const hash = type(/^[A-Fa-f0-9]{64}$/, "@", "hash");
```

Comment:
We shouldn't need a hash type anymore. If we ever need a hash, we should be able to use a Cid instead.